cmd/gpu-kubelet-plugin/device_health.go

```diff
@@ -0,0 +1,342 @@
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"k8s.io/klog/v2"
+)
+
+const (
+	FullGPUInstanceID uint32 = 0xFFFFFFFF
+)
+
+// For a MIG device the placement is defined by the 3-tuple <parent UUID, GI, CI>.
+// For a full device the returned 3-tuple is the device's uuid and (FullGPUInstanceID) 0xFFFFFFFF for the other two elements.
+type devicePlacementMap map[string]map[uint32]map[uint32]*AllocatableDevice
+
+type nvmlDeviceHealthMonitor struct {
+	nvmllib           nvml.Interface
+	eventSet          nvml.EventSet
+	unhealthy         chan *AllocatableDevice
+	cancelContext     context.CancelFunc
+	deviceByPlacement devicePlacementMap
+	skippedXids       map[uint64]bool
+	wg                sync.WaitGroup
+}
+
+func newNvmlDeviceHealthMonitor(config *Config, allocatable AllocatableDevices, nvdevlib *deviceLib) (*nvmlDeviceHealthMonitor, error) {
+	if nvdevlib.nvmllib == nil {
+		return nil, fmt.Errorf("nvml library is nil")
+	}
+	if ret := nvdevlib.nvmllib.Init(); ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("failed to initialize NVML: %v", ret)
+	}
+	defer func() {
+		_ = nvdevlib.nvmllib.Shutdown()
+	}()
+
+	m := &nvmlDeviceHealthMonitor{
+		nvmllib:           nvdevlib.nvmllib,
+		unhealthy:         make(chan *AllocatableDevice, len(allocatable)),
+		deviceByPlacement: getDevicePlacementMap(allocatable),
+		skippedXids:       xidsToSkip(config.flags.additionalXidsToIgnore),
+	}
+	return m, nil
+}
+
+func (m *nvmlDeviceHealthMonitor) Start(ctx context.Context) (rerr error) {
+	if ret := m.nvmllib.Init(); ret != nvml.SUCCESS {
+		return fmt.Errorf("failed to initialize NVML: %v", ret)
+	}
+
+	defer func() {
+		if rerr != nil {
+			_ = m.nvmllib.Shutdown()
+		}
+	}()
+
+	klog.V(6).Info("creating NVML events for device health monitor")
+	eventSet, ret := m.nvmllib.EventSetCreate()
+	if ret != nvml.SUCCESS {
+		return fmt.Errorf("failed to create event set: %w", ret)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	m.cancelContext = cancel
+
+	m.eventSet = eventSet
+
+	klog.V(6).Info("registering NVML events for device health monitor")
+	m.registerEventsForDevices()
+
+	klog.V(6).Info("started device health monitoring")
+	m.wg.Add(1)
+	go m.run(ctx)
+
+	return nil
+}
+
+func (m *nvmlDeviceHealthMonitor) registerEventsForDevices() {
+	eventMask := uint64(nvml.EventTypeXidCriticalError | nvml.EventTypeDoubleBitEccError | nvml.EventTypeSingleBitEccError)
+
+	for parentUUID, giMap := range m.deviceByPlacement {
+		gpu, ret := m.nvmllib.DeviceGetHandleByUUID(parentUUID)
+		if ret != nvml.SUCCESS {
+			klog.Infof("Unable to get device handle from UUID[%s]: %v; marking it as unhealthy", parentUUID, ret)
+			m.markAllSlicesUnhealthy(giMap)
+			continue
+		}
+
+		supportedEvents, ret := gpu.GetSupportedEventTypes()
+		if ret != nvml.SUCCESS {
+			klog.Infof("unable to determine the supported events for %s: %v; marking it as unhealthy", parentUUID, ret)
+			m.markAllSlicesUnhealthy(giMap)
```

Comment:
same questions/argument as in https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/689/files#r2577289361